### PR TITLE
Add pyvista plot directive

### LINKED ISSUE #1426: Capture Plotting in docstring examples
The following scipy example has embedded a matplotlib into a docstring page.  There are many method docstring where we issue (and suppress) plotting in the example section.  Would be awesome if we could implement the same for pyvista.

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.oaconvolve.html#scipy.signal.oaconvolve

### PR DESCRIPTION
**Pull request by akaszynski**
##### tl;dr
Resolves #1426 by adding a new sphinx directive that automatically captures and renders any plots in our example sections. 

### Add plot directive and sphinx extension

Turns out there's an automated way of generating plots from our example sections, and it means that we can now automatically generating plots directly from our examples without having to skip them.  Since we're moving in the direction of one page from method/function, this is leading to some really clear examples:

tmp

We can also generate figures using the ``.. pyvista-plot::`` directive within our docs rather than using `jupyter-execute`.  This is multi-thread compatible and has none of the issues where the jupyter plots fail to come up in html.

### Notes
- This uses much of the code originally developed in https://github.com/matplotlib/matplotlib/raw/master/lib/matplotlib/sphinxext/plot_directive.py.  Without it, I would have never been able to figure this out.
- We directly monkey patch `sphinx.GoogleDocstring._parse_generic_section` in our `conf.py`.  There wasn't a way around this.  This is why `numpydoc` is the only plotter that actually allows you to automatically display `matplotlib` plots in their docs; they own the documentation tool.
- This PR disables outputting `cpos` when issuing `show`.  It's leading to messy code and the output from a `show` plot should be the plot when looking at examples.
- Have fun @adeak, this is a long one.

### CODE REVIEW

**Comment by akaszynski** [comment]
Pinging @pyvista/developers.

Ready for review/comments.  It's fairly major and I'd like opinions of other developers.

File: pyvista/plotting/camera.py [hunk 1]
```diff
@@ -150,11 +150,6 @@ def model_transform_matrix(self, matrix):
         vtk_matrix.DeepCopy(matrix.ravel())
         self.SetModelTransformMatrix(vtk_matrix)
 
-    @property
-    def is_parallel_projection(self):
-        """Return True if parallel projection is set."""
-        return self._is_parallel_projection
-
```
Comment (tkoyama010) [inline pyvista/plotting/camera.py:157]: How about raise [DeprecationWarning](https://docs.python.org/3/library/exceptions.html#DeprecationWarning) about `is_parallel_projection` ?

File: doc/conf.py [hunk 2]
```diff
@@ -140,6 +145,48 @@
...
+            else:
+                lines.append(line)
+
+    if lines:
+        return [header, ''] + lines + ['']
+    else:
+        return [header, '']
+
+# override method
+GoogleDocstring._parse_generic_section = _custom_parse_generic_section
```
Comment (larsoner) [inline doc/conf.py:187]: Playing with private variables is dangerous / fragile. Did you consider doing whatever `matplotlib` does with their directive instead?

https://github.com/matplotlib/matplotlib/blob/c06e8709dde6504d396349c0c80ef019c88c3927/lib/matplotlib/sphinxext/plot_directive.py

Also the current implementation relies on `napoleon`, so will only work for people who use that to document their code (and not, say`numpydoc`).

File: tests/test_plot_directive.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+"""
+
+"""
```
Comment (tkoyama010) [inline tests/test_plot_directive.py:3]: Is the blank test intended?

Comment (akaszynski) [reply tests/test_plot_directive.py:3]: This file should be deleted.  Plot directive is checked in test_tinypages.py

File: ignore_words.txt [hunk 4]
```diff
@@ -2,3 +2,4 @@ lod
 byteorder
 flem
 parm
+tht
```
Comment (adeak) [inline ignore_words.txt:5]: I can't `git grep tht` anywhere. So why the ignore?

File: doc/conf.py (hunk 2 above)
Comment (akaszynski) [reply doc/conf.py:187]: We do exactly what matplotlib did for their directive, it's basically a carbon copy with changes to scrape screenshots rather than matplotlib figures.

This part here is for automating the addition of the rubric when parsing the Examples sections.  Effectively, when it detects that pyvista is used in an examples section, it adds the `pyvista-plot` directive to the section.  Numpydoc does the same thing, which is why you can't use Napoleon and scrape the matplotlib figures; it's literally baked into their source.

As far as I can tell, there's no way of hooking into the section parser.  We could ask for them to add a config value or hook for this, which would make the code much "safer"; I agree that this is a brittle approach.

File: pyvista/plotting/camera.py (hunk 1 above)
Comment (akaszynski) [reply pyvista/plotting/camera.py:157]: Yes, could you add that?

File: pyvista/ext/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+"""Module level init for ``pyvista.ext``."""
```
Comment (adeak) [inline pyvista/ext/__init__.py:1]: Shouldn't this be either less, or more?

File: ignore_words.txt (hunk 4 above)
Comment (akaszynski) [reply ignore_words.txt:5]: That's my fault; I added this as my rst checker compiles on the fly and there are some temporary files that are being flagged by the spell check.  We can remove this.

File: doc/conf.py (hunk 2 above)
Comment (larsoner) [reply doc/conf.py:187]: Ahh okay, makes sense. +1 for opening an issue with the napolean folks. If they could add some sort of post-processing hook into the examples section that would be great!

File: tests/test_polydata.py [hunk 6]
```diff
@@ -268,7 +268,8 @@ def test_multi_ray_trace(sphere):
 @pytest.mark.skipif(not system_supports_plotting(), reason="Requires system to support plotting")
 def test_plot_curvature(sphere):
     cpos = sphere.plot_curvature(off_screen=True)
-    assert isinstance(cpos, pyvista.CameraPosition)
+    # cpos is None within unit testing
+    assert cpos is None
```
Comment (adeak) [inline tests/test_polydata.py:272]: I've only barely skimmed the PR yet so this might be answered elsewhere: is it specified when `cpos` is returned and when it isn't?

And is it not a bad idea to change API behaviour for testing? When we're no longer testing production code, tests start losing meaning.

If I understand correctly the goal is to make example code blocks pretty by suppressing `cpos` being returned. So can't we monkeypatch `plot`/`show` /etc. calls only for docs generation? This shouldn't affect unit tests.

File: tests/tinypages/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+_build/
```
Comment (adeak) [inline tests/tinypages/.gitignore:1]: Does the example project have to be a git repo? If it doesn't/it isn't, does it need a `.gitignore`?

And what about `tinypages/_static` with an empty `.gitignore`?

File: ignore_words.txt (hunk 4 above)
Comment (tkoyama010) [reply ignore_words.txt:5]:
```suggestion
```

File: pyvista/plotting/camera.py (hunk 1 above)
Comment (tkoyama010) [reply pyvista/plotting/camera.py]: Done in https://github.com/pyvista/pyvista/pull/1457/commits/0b2a8cf0664ac0f89f621d4da42bc71593cda00b

**Review by tkoyama010 — APPROVED** [review]
LGTM after all conversation is resolved.

File: doc/conf.py (hunk 2 above)
Comment (akaszynski) [reply doc/conf.py:187]: Done in https://github.com/sphinx-contrib/napoleon/issues/31.

File: pyvista/ext/__init__.py (hunk 5 above)
Comment (akaszynski) [reply pyvista/ext/__init__.py:1]: Our docs complain about not having summary line, so I added one.  Could add more, but this matches what we have in our other init modules.

File: tests/test_polydata.py (hunk 6 above)
Comment (akaszynski) [reply tests/test_polydata.py:272]: I honestly would love to remove returning `cpos` altogether and then add an optional `return_cpos` to our kwargs.  How about we do that instead.

I agree that it's bad practice to change outputs just for testing, which defeats testing...

Comment (adeak) [reply tests/test_polydata.py:272]: A new kwarg would work. Then again you've been working to make return values homogeneous for `inplace=True`, so this would then defeat _that_ a bit. But practicality beats purity so it could work.

File: tests/tinypages/.gitignore (hunk 7 above)
Comment (akaszynski) [reply tests/tinypages/.gitignore:1]: I'm removing this.  I quite literally copied and pasted this out of matplotlib.

File: tests/test_polydata.py (hunk 6 above)
Comment (akaszynski) [reply tests/test_polydata.py:272]: If you're fine with it for now, let's keep as is and fix in a follow-up PR.

Comment (adeak) [reply tests/test_polydata.py:272]: Well I'm not crazy about it, but if you promise to fix that soon in the follow-up then it won't be like this for long :P

Comment (akaszynski) [reply tests/test_polydata.py:272]: > Well I'm not crazy about it, but if you promise to fix that soon in the follow-up then it won't be like this for long :P

This PR is huge as is.  I'll make another PR correcting the behavior.

**Comment by adeak** [comment]
@tkoyama010 I'd like to ask that next time when there's such a huge PR we either wait a few days or explicitly ask before merging. Normally there's no rush and contributor workloads vary. If there's a heads-up that someone's about to merge a PR we can either say it's fine or try to make time to look at it sooner.

If that helps I can try paying special attention to formally ask reviews from myself on PRs that I still want to look at.

**Comment by tkoyama010** [comment]
@adeak Thank you for your comment. In the case of major revisions, I will set aside enough time for review.
